### PR TITLE
Upgrade to Plexil 6

### PR DESCRIPTION
**Pull request by kmdalal**
### Summary
- The Plexil interface adapter is reworked to use the new Plexil 6 API.  A variety of new classes and methods are used, and Plexil commands and lookups are registered using the new API.
- Misc
  - Some formerly static file variables are now class members.
  - The `TiltAndImagePass` plan had a needed tilt command restored.
  - The `ArmEndEffectorFT` lookup was added (it was forgotten in earlier works).
  - Some minor unrelated refactoring of class methods.
  - Cleanup of header file inclusions, with unneeded and redundant ones removed.
  - Minor code reformatting, reorganization, and revision of comments.

### Reviewer Suggestions
- @AstroStucky : OceanWATERS and OWLAT builds and tests, code review.
- @Keeganfn : OceanWATERS build and test, code review.

Note that many of the code changes will not be understandable without knowing the PLEXIL API.  I suggest looking for consistency, and relying on the tests for correct operation.

### Install Plexil 6

This is a prerequisite for testing.  Install it so that your existing Plexil (4.6) can still be used.  This is done simply by resetting the `PLEXIL_HOME` environment variable.  Do not delete or overwrite your existing Plexil installation!

1. cd to a directory where you wish to install Plexil 6 (alongside your existing Plexil installation is reasonable).
2. Clone it.
    `git clone -b releases/plexil-6 https://github.com/plexil-group/plexil.git plexil-6`
3. Reset  `PLEXIL_HOME`, putting the following line in your shell init file (e.g. `.bashrc`) and using the correct path:
    `export PLEXIL_HOME=/path/to/plexil-6`
4. Configure for the build:
    ```
    cd $PLEXIL_HOME
    git submodule update --init --recursive
    make src/configure
    cd src
    ./configure CFLAGS="-g -O2" CXXFLAGS="-g -O2" --prefix=$PLEXIL_HOME --disable-static --disable-viewer --enable-ipc
    ```
5. Build PLEXIL.
    ```
    cd $PLEXIL_HOME
    make universalExec plexil-compiler checkpoint
    ```

### Build and test OceanWATERS

1. With all other repositories on the `noetic-devel` branch, do a `catkin clean` followed by `catkin build`.
2. Start the Atacama world.
3. In a separate shell, start the `ow_exec` launch.
4. Run the plan `TestTelemetry.plx`.  Output should approximately like the following log:
[ow-test-telemetry.log](https://github.com/nasa/ow_autonomy/files/12384577/ow-test-telemetry.log)
5. Run the plan `TestActions.plx`.  It should run without errors and the output should look like the following log:
[ow-test-actions.log](https://github.com/nasa/ow_autonomy/files/12384648/ow-test-actions.log)

### Build and test OWLAT

1. With all other repositories on the `noetic-devel` branch, do a `catkin clean` followed by `catkin build -DOWLAT=ON`.
2. Start `roscore` by typing just this in a shell.
3. In a separate shell start the OWLAT autonomy interface:
    ```
     source /opt/ros/noetic/setup.bash
     roslaunch owl_msgs autonomy_interface.launch
4. In a separate shell, start a Python3 virtual environment, and then start OWLAT:
    ```
    launch_owlat Scoop CAM_FORCE_MOCK:true PT_USE_MOCK:true
    ```
5. In a separate shell, start the `owlat_exec` launch:
    ```
    roslaunch ow_plexil owlat_exec.launch
    ```
6. Run the plan `TestTelemetry.plx`.  Output should approximately like the following log:
[owlat-test-telemetry.log](https://github.com/nasa/ow_autonomy/files/12384574/owlat-test-telemetry.log)
7. Run the plan `TestOwlatActions.plx`.  It may take over 20 minutes to complete.  It should run without errors and the output should look like the following log:
[owlat-test-actions.log](https://github.com/nasa/ow_autonomy/files/12384576/owlat-test-actions.log)

### CODE REVIEW

**Review by Keeganfn — APPROVED** [review]
All the tests pass and the code looks good. The lookup interface is much better looking now! Thanks!

**Comment by AstroStucky** [comment]
I keep getting linkage errors when I try to compile plexil-6

[plexil-6-make.log](https://github.com/nasa/ow_autonomy/files/12432783/plexil-6-make.log)

**Review by AstroStucky — CHANGES_REQUESTED** [review]

**Comment by AstroStucky** [comment]
My plexil-6 build issue resolved itself. After a clean build of plexil-6 everything worked fine, so I assume I missed a step along the way.

**Comment by AstroStucky** [comment]
Step 4 of testing OW I see `nan` values for battery telemetry
```
[ INFO:/ow_exec_node] PLEXIL: Battery state of charge: nan
[ INFO:/ow_exec_node] PLEXIL: Battery remaining useful life: nan
[ INFO:/ow_exec_node] PLEXIL: Battery temperature: nan
```
and this may be an already known error, but the "Enter any additional" plan prompt never showed up when it concluded, and I could not type in the next plan.

**Comment by AstroStucky** [comment]
Step 5 I consider a passed, although I did notice the usual problems that I think were fixed in branch OCEANWATER-1102_dig_force_aware_procedures and OCEANWATER-1154_update_action_tests_for_new_relative_flag_behavior. 

I do not recall why we have not merged them in yet. I am in the process of the merging in releases/release-11-1 into them both because they are now far behind.

**Comment by AstroStucky** [comment]
> Step 5 I consider a passed, although I did notice the usual problems that I think were fixed in branch OCEANWATER-1102_dig_force_aware_procedures and OCEANWATER-1154_update_action_tests_for_new_relative_flag_behavior.
> 
> ~I do not recall why we have not merged them in yet. I am in the process of the merging in releases/release-11-1 into them both because they are now far behind.~

**Corrections:** 

- OCEANWATER-1154_update_action_tests_for_new_relative_flag_behavior was merged into noetic-devel, but I believe its fixes were never brought to the releases/release-11-1 branch. 
- OCEANWATER-1102_dig_force_aware_procedures probably should be merged at the same time as OCEANWATER-1146_new_dig_model_integration

**Comment by kmdalal** [comment]
> Step 4 of testing OW I see `nan` values for battery telemetry
> 
> ```
> [ INFO:/ow_exec_node] PLEXIL: Battery state of charge: nan
> [ INFO:/ow_exec_node] PLEXIL: Battery remaining useful life: nan
> [ INFO:/ow_exec_node] PLEXIL: Battery temperature: nan
> ```

This is expected for OWLAT but not OceanWATERS.  Are the other repos on the 'master' branch (that actually shouldn't matter here)?  Was the simulator fully started before you ran this plan?

> and this may be an already known error, but the "Enter any additional" plan prompt never showed up when it concluded, and I could not type in the next plan.

That suggests the `TestTelemetry` plan had not yet finished, which is odd, but could happen if the simulator wasn't fully started and a Lookup never returned.

Please see if this problem is reproduceable, unless you already know it is.

**Comment by AstroStucky** [comment]
I was not able to reproduce the nans, so I suspect you are correct that I did not give the simulator enough time to start. 

However, the plan prompt still did not come up after the completion of TestTelemetry.plx

**Comment by AstroStucky** [comment]
My OWLAT test only got as far as 
```
[ INFO:/owlat_exec_node] PLEXIL: Finished ArmFindSurface test.
[ INFO:/owlat_exec_node] PLEXIL: Starting TaskPSP test...
[ INFO:/owlat_exec_node] ArmUnstow () started...
[ INFO:/owlat_exec_node] ArmUnstow finished in state SUCCEEDED
```
And it froze. The clock in DARTS is no longer ticking, and I cannot move the camera. There is no error output, everything just stopped. 

I will reattempt the test. I suspect it is a problem with DARTS and not this PR.

**Comment by AstroStucky** [comment]
It turns out it froze when the tool was changed. My second test has not frozen so far.

**Comment by AstroStucky** [comment]
OWLAT Tests passed!

File: ow_plexil/src/plexil-adapter/OwAdapter.cpp [hunk 1]
```diff
@@ -353,53 +282,89 @@ static void identify_sample_location (Command* cmd, AdapterExecInterface* intf)
...
 
-  if (! lookup(state.name(), state.parameters(), retval)) {
-    ROS_ERROR("PLEXIL Adapter: Invalid lookup name: %s", state.name().c_str());
-  }
-  entry.update(retval);
+  debugMsg("OwAdapter", " initialized.");
+  return LanderAdapter::initialize (config);
 }
 
 extern "C" {
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/OwAdapter.cpp:436]: Actually the children (leaf classes) are the _only_ place this should occur.  These are the specific adapters being registered.

File: ow_plexil/src/plexil-adapter/LanderAdapter.cpp [hunk 2]
```diff
@@ -18,12 +18,8 @@ using std::string;
 using std::unique_ptr;
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/LanderAdapter.cpp:18]: Good catch.  I've moved all `using` statements after all includes in all the files where needed.

File: ow_plexil/src/plexil-adapter/LanderAdapter.h [hunk 3]
```diff
@@ -19,27 +21,31 @@ class LanderAdapter : public PlexilAdapter
...
   static float TiltMinDegrees;
   static float TiltMaxDegrees;
   static float PanTiltInputTolerance;
 
-  // No default constructor, only this specialized one.
-  LanderAdapter (PLEXIL::AdapterExecInterface&, const pugi::xml_node&);
-  virtual ~LanderAdapter () = default;
+  LanderAdapter (PLEXIL::AdapterExecInterface&, PLEXIL::AdapterConf*);
+  LanderAdapter () = delete;
+  virtual ~LanderAdapter () = 0;
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/LanderAdapter.h:36]: If you meant the destructor, it is defaulted, in the .cpp file.  One can't both default and make pure virtual inside the class, and I avoid extra definitions besides the class in the header file just as a stylistic preference.

File: ow_plexil/src/plexil-adapter/OwExecutive.cpp [hunk 4]
```diff
@@ -30,11 +10,31 @@ using PLEXIL::InterfaceSchema;
...
+#include <AdapterExecInterface.hh>
+#include <Debug.hh>
+#include <Error.hh>
+#include <InterfaceSchema.hh>
+#include <parsePlan.hh>
+#include <State.hh>
+using namespace PLEXIL;
+
+// OW
+#include "OwExecutive.h"
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/OwExecutive.cpp:29]: Done.

File: ow_plexil/src/plexil-adapter/PlexilAdapter.cpp [hunk 5]
```diff
@@ -14,25 +16,33 @@
...
+static void angles_equivalent (const State& s, LookupReceiver* r)
+{
+  double deg1, deg2, tolerance;
+  s.parameters()[0].getValue(deg1);
+  s.parameters()[1].getValue(deg2);
+  s.parameters()[2].getValue(tolerance);
+  r->update(PlexilInterface::anglesEquivalent(deg1, deg2, tolerance));
+}
+
+PlexilAdapter::~PlexilAdapter() = default;
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilAdapter.cpp:32]: Same reply as to the same thing in another file -- it's just a stylistic preference over having extra definitions outside the class in the header file (when inlining doesn't buy much).

File: ow_plexil/src/plexil-adapter/PlexilInterface.h [hunk 6]
```diff
@@ -18,6 +18,8 @@ constexpr double R2D = 180.0 / M_PI ;
 class PlexilInterface
 {
  public:
+  static bool anglesEquivalent (double deg1, double deg2, double tolerance);
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilInterface.h:21]: I don't understand the question -- this outputs a boolean.  If you mean the range of accepted inputs, I think it should be (-infinity, +infinity), with tolerance >= 0.  Not sure how valuable this information is, but I can document it in comments...

This is an old function that was only moved to another file for this PR.

File: ow_plexil/src/plexil-adapter/PlexilAdapter.h [hunk 7]
```diff
@@ -8,37 +8,33 @@
...
-  virtual bool shutdown();
-  virtual void invokeAbort(PLEXIL::Command *cmd);
-  virtual void subscribe(const PLEXIL::State& state);
-  virtual void unsubscribe(const PLEXIL::State& state);
-  virtual void lookupNow (const PLEXIL::State&, PLEXIL::StateCacheEntry&) { }
+  virtual bool initialize(PLEXIL::AdapterConfiguration*) override;
+  virtual bool start() override;
+  virtual void stop() override;
+  virtual void subscribe(const PLEXIL::State& state); // obsolete?
+  virtual void unsubscribe(const PLEXIL::State& state);  // obsolete?
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilAdapter.h:32]: Good catch.  Removed.

**Comment by kmdalal** [comment]
> However, the plan prompt still did not come up after the completion of TestTelemetry.plx

This happens for me sometimes.  It's a known issue, not related to this PR, involving plans that finish quickly.  Even though the prompt is not there, you can enter the next plan name and it will execute.  

I'll create a Jira for this issue and make a note about it in the user documentation.

**Comment by kmdalal** [comment]
I've responded to all comments left so far and pushed another round of commits.  Let me know if I missed anything.

File: ow_plexil/src/plexil-adapter/LanderAdapter.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (AstroStucky) [reply ow_plexil/src/plexil-adapter/LanderAdapter.h:36]: You are correct. This is just a peculiar C++ use case I have not seen before.

File: ow_plexil/src/plexil-adapter/PlexilInterface.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (AstroStucky) [reply ow_plexil/src/plexil-adapter/PlexilInterface.h:21]: My mistake, this comment should have been for a different function in different file. Not sure how I put it here...

File: ow_plexil/src/plexil-adapter/PlexilInterface.cpp [hunk 8]
```diff
@@ -10,11 +10,23 @@ using std::string;
 using std::map;
 using std::make_unique;
 
-// Dummy operation ID that signifies idle lander operation.
-#define IDLE_ID (-1)
+///////////////////////// Utilities /////////////////////////
+
+static double normalize_degrees (double angle)
```
Comment (AstroStucky) [inline ow_plexil/src/plexil-adapter/PlexilInterface.cpp:15]: My range comment was intended for this function. Do you know what ranges of angles this results in? Can you put it in a comment?

**Review by AstroStucky — COMMENTED** [review]
My range comment was confusing because it was made on the wrong line in the wrong file. Please take a look at the updated version.

File: ow_plexil/src/plexil-adapter/PlexilInterface.cpp (hunk 8 above)
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/PlexilInterface.cpp:15]: I'd have to study it, and this function has been around long before this PR.  I've made a to-do to add the comment for Release 12 (comments don't need PRs) but suggest leaving it out of this review.

**Review by AstroStucky — APPROVED** [review]
Retested, and everything worked. Great work! Glad to see the final commit of 11.1

**Pull request merged**